src/Orleans.Transactions/State/ReaderWriterLock.cs

```diff
@@ -252,6 +160,201 @@ private Task LockWork()
             return task;
         }
 
+        public bool ValidateLock(Guid transactionId, AccessCounter accessCount, out TransactionalStatus status, out TransactionRecord<TState> record)
+        {
+            if (currentGroup == null || !currentGroup.TryGetValue(transactionId, out record))
+            {
+                record = new TransactionRecord<TState>()
+                {
+                    TransactionId = transactionId
+                };
+                status = TransactionalStatus.BrokenLock;
+                return false;
+            }
+            else if (record.NumberReads != accessCount.Reads
+                   || record.NumberWrites != accessCount.Writes)
+            {
+                Rollback(transactionId, true);
```

Comment:
We dropped a message here - maybe at some point in future we should add Trace-level logging here instead.